atlasaction/action.go

```diff
@@ -383,6 +385,133 @@ func (a *Actions) SchemaTest(ctx context.Context) error {
+			// The user did not provide a name.
+			// Generate unique name base on the PR number
+			params.Name = fmt.Sprintf("pr-%d-%.8s", tc.PullRequest.Number, plan.File.ToHash)
+		}
+		planURL, err := url.JoinPath(params.Repo, "plans", params.Name)
+		if err != nil {
+			return fmt.Errorf("failed to generate plan URL: %w", err)
+		}
+		// Check if the plan already exists
+		switch _, err = a.Atlas.SchemaPlanPull(ctx, &atlasexec.SchemaPlanPullParams{
+			ConfigURL: params.ConfigURL,
+			Env:       params.Env,
+			Vars:      params.Vars,
+			URL:       planURL,
+		}); {
+		// There is a plan
+		case err == nil:
+			a.Infof("Schema plan already exists, linting the plan %q", params.Name)
+			plan, err = a.Atlas.SchemaPlanLint(ctx, &atlasexec.SchemaPlanLintParams{
+				ConfigURL: params.ConfigURL,
+				Env:       params.Env,
+				Vars:      params.Vars,
+				Context:   params.Context,
+				DevURL:    params.DevURL,
+				From:      params.From,
+				To:        params.To,
+				Repo:      params.Repo,
+				File:      planURL,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to get the schema plan: %w", err)
+			}
+			// FIXME(giautm): Remove this workaround after fixing the schema URL issue.
+			plan.File.URL = planURL
+		// The plan does not exist
+		case strings.Contains(err.Error(), fmt.Sprintf("plan %q was not found", params.Name)):
+			a.Infof("Schema plan does not exist, creating a new one with name %q", params.Name)
+			params.DryRun = false // Disable dry-run
+			params.Pending = true // Submit as pending
+			plan, err = a.Atlas.SchemaPlan(ctx, params)
+			if err != nil {
+				return fmt.Errorf("failed to save schema plan: %w", err)
+			}
+		// Unknown error
+		default:
+			return err
+		}
+		summary, err := schemaPlanComment(plan, params.Env)
+		if err != nil {
+			return fmt.Errorf("failed to generate schema plan comment: %w", err)
+		}
+		a.AddStepSummary(summary)
+		// Comment on the PR
+		err = a.GithubClient(tc.Repo, tc.SCM.APIURL).
+			UpsertComment(ctx, tc.PullRequest, plan.File.FromHash, summary)
+		if err != nil {
+			a.Errorf("failed to comment on the pull request: %v", err)
+		}
+		return nil
+	default:
+		if params.Name == "" {
+			// The user did not provide a name.
+			// Generate unique name base on the PR number
+			c := a.GithubClient(tc.Repo, tc.SCM.APIURL)
+			switch pr, err := c.MergedPullRequest(ctx, tc.Commit); {
+			case err != nil:
+				return fmt.Errorf("failed to get the merged pull request: %w", err)
+			case pr != nil:
+				params.Name = fmt.Sprintf("pr-%d-%.8s", pr.Number, plan.File.ToHash)
+			default:
+				a.Infof("No merged pull request found for commit %q, skip the approval", tc.Commit)
+				return nil
+			}
+		}
+		planURL, err := url.JoinPath(params.Repo, "plans", params.Name)
+		if err != nil {
+			return fmt.Errorf("failed to generate plan URL: %w", err)
+		}
+		// Approve the plan if it exist.
+		switch result, err := a.Atlas.SchemaPlanApprove(ctx, &atlasexec.SchemaPlanApproveParams{
```

Comment:
I think there should be a separate action for approving. Let the user control the approval instead of approving automatically based on the conditions the action ran in.
cc @a8m.